optimum/neuron/modeling_seq2seq.py

```diff
@@ -0,0 +1,912 @@
+
+            # Cases that can be handled in XLA without requiring
+            # non-padded input_ids
+            if isinstance(stopping_criteria, MaxLengthCriteria):
+                stop_criterion_2 = seq_length >= stopping_criteria.max_length
+            elif isinstance(stopping_criteria, MaxTimeCriteria):
+                stop_criterion_2 = stopping_criteria(input_ids, scores)
+            else:
+                # Other cases will be handled on CPU
+                batch_size, _ = input_ids.shape
+                mask = torch.cat(
+                    [torch.ones(batch_size, seq_length), torch.zeros(batch_size, input_ids.shape[1] - seq_length)],
+                    dim=1,
+                ).bool()
+                input_ids_cpu = torch.masked_select(input_ids, mask).reshape((batch_size, seq_length)).to("cpu")
+                scores_cpu = scores.to("cpu") if torch.is_tensor(scores) else scores
+                stop_criterion_2 = stopping_criteria(input_ids_cpu, scores_cpu)
+
+            if stop_criterion_1 or stop_criterion_2:
+                this_peer_finished = True
+
+            if this_peer_finished:
+                break
+
+        if streamer is not None:
+            streamer.end()
+
+        return input_ids
+
+    def _reorder_cache(self, past_key_values, beam_idx):
+        """
+        The cache was reordered during the tracing of the decoder so we can skip it here. This is needed for beam search and not greedy sampling.
+        """
+        self.beam_idx = beam_idx
+        return past_key_values
+
+    def get_encoder(self) -> "NeuronEncoder":
+        return self.encoder
+
+    def _update_model_kwargs_for_xla_generation(
+        self,
+        model_kwargs: Dict[str, Any],
+        batch_size: int,
+        is_encoder_decoder: bool = False,
+        standardize_cache_format: bool = False,
+        max_length: Optional[int] = None,
+        seq_length: Optional[int] = None,
+        use_cache: bool = True,
+    ) -> Dict[str, Any]:
+        def _update_attention(model_kwargs, is_encoder_decoder):
+            """Updates the appropriate attention mask -- encoder-decoder models use `decoder_attention_mask`"""
+
+            attention_mask_name = "decoder_attention_mask" if is_encoder_decoder else "attention_mask"
+            attention_mask = model_kwargs.pop(attention_mask_name)
+            attention_mask_update_slice = torch.ones(
+                (batch_size, 1), dtype=attention_mask.dtype, device=attention_mask.device
+            )
+            attention_mask = torch.cat([attention_mask[:, 1:], attention_mask_update_slice], dim=-1)
+            mask = {attention_mask_name: attention_mask}
+            return mask
+
+        mask = _update_attention(model_kwargs, is_encoder_decoder)
+        # sets the updated variables (mask and past_key_values)
+        model_kwargs.update(mask)
+
+        # Set a mock cache tensor
+        model_kwargs["past_key_values"] = torch.tensor([])
```

Comment:
Why do you need this, since you override both generation loops and ignore that parameter completely ?

Reply:
Nope, I override the inputs preparation func, will remove it.

Reply:
Good catch!